### PR TITLE
Custom sorting for items

### PR DESCRIPTION
**Pull request by bhollis**
Wow, this has been a long time coming - it's one of our oldest open issues, #647.

This PR implements a new, React-based sorting configuration widget, plus a collection of higher-order functions for building custom comparators, and uses them together to implement custom sorting. People can now choose which properties they want to sort by, and in which order. The whole thing can be controlled by buttons or drag and drop.

custom-sort

With this in place we can add new custom sorting options much more easily. D2 has a lot fewer interesting things to sort on (archetype, maybe?) but that might still be interesting. We can also theoretically get rid of some of the "preset" options - **any suggestions**? I'm tempted to drop the "item type then primary stat" ones. If we do, we'll just transparently change it into a "custom" variant.

As a side effect of switching to these new sorting methods, I think we render the inventory a lot faster, too.

This same component can be used for https://github.com/DestinyItemManager/DIM/issues/208, too.

### CODE REVIEW

**Comment by duckmanBR** [comment]
It's possible to give the option to select, for each option, if you want an ascending/descending sort?

**Comment by bhollis** [comment]
I thought about it, but I’m not sure it’s that valuable. I couldn’t think of a case where it’d be really helpful. Most of these have an obvious good/bad direction.

**Comment by SunburnedGoose** [comment]
Let's trim down the prebuilt list, and push people towards this component. If the prebuilt checkboxes were like presets maybe, that change the component, it could double as education to teach players how to use it...

**Comment by bhollis** [comment]
Cool. Any suggestions on which to keep? They do act like presets, but I didn’t have them overwrite the custom sort in case folks have a separate custom sort option that they want to keep in place.

File: src/app/settings/settings.html [hunk 1]
```diff
@@ -67,35 +67,12 @@ <h2 ng-i18next="Settings.Items"></h2>
...
-            <span ng-i18next="Settings.SortName"></span>
-          </label>
-          <label>
-            <input type="radio" ng-model="vm.settings.itemSort" value="typeThenPrimary">
-            <span ng-i18next="Settings.SortTypeThenPrimary"></span>
-          </label>
-          <label>
-            <input type="radio" ng-model="vm.settings.itemSort" value="typeThenName">
-            <span ng-i18next="Settings.SortTypeThenName"></span>
+          <label ng-repeat="(value, i18key) in vm.itemSortPresets track by value">
```
Comment (delphiactual) [inline src/app/settings/settings.html]: Minor nit: i18nkey

File: src/app/settings/settings.html [hunk 2]
```diff
@@ -67,35 +67,12 @@ <h2 ng-i18next="Settings.Items"></h2>
...
+            <input type="radio" ng-model="vm.settings.itemSort" value="{{value}}">
+            <span ng-i18next="{{i18key}}"></span>
```
Comment (delphiactual) [inline src/app/settings/settings.html]: same here

**Comment by bhollis** [comment]
I chose to remove the "base power", "item type then primary" and "item type by name" presets.

**Comment by kyleshay** [comment]
wonder if we should add a sort option on the main page, so users can quickly change sorts around while organizing their items? maybe a button next to the vault character card?

**Comment by bhollis** [comment]
I’ve thought about it. I’m a bit hesitant because I think people use sorting when they should use searches or the compare tool.

**Pull request merged**